xarray/core/indexes.py

```diff
@@ -45,72 +88,315 @@ def concat(
-        pandas.Index object.
+        This method is used by all xarray operations that still rely on
+        pandas.Index objects.
 
+        By default it raises a ``TypeError``, unless it is re-implemented in
+        subclasses of Index.
         """
         raise TypeError(f"{self!r} cannot be cast to a pandas.Index object")
 
     def isel(
-        self, indexers: Mapping[Any, int | slice | np.ndarray | Variable]
-    ) -> Index | None:
+        self: T_Index, indexers: Mapping[Any, int | slice | np.ndarray | Variable]
+    ) -> T_Index | None:
+        """Maybe returns a new index from the current index itself indexed by
+        positional indexers.
+
+        This method should be re-implemented in subclasses of Index if the
+        wrapped index structure supports indexing operations. For example,
+        indexing a ``pandas.Index`` is pretty straightforward as it behaves very
+        much like an array. By contrast, it may be harder doing so for a
+        structure like a kd-tree that differs much from a simple array.
+
+        If not re-implemented in subclasses of Index, this method returns
+        ``None``, i.e., calling :py:meth:`Dataset.isel` will either drop the
+        index in the resulting dataset or pass it unchanged if its corresponding
+        coordinate(s) are not indexed.
+
+        Parameters
+        ----------
+        indexers : dict
+            A dictionary of positional indexers as passed from
+            :py:meth:`Dataset.isel` and where the entries have been filtered
+            for the current index.
+
+        Returns
+        -------
+        maybe_index : Index
+            A new Index object or ``None``.
+        """
         return None
 
     def sel(self, labels: dict[Any, Any]) -> IndexSelResult:
+        """Query the index with arbitrary coordinate label indexers.
+
+        Implementation is optional but required in order to support label-based
+        selection. Otherwise it will raise an error when trying to call
+        :py:meth:`Dataset.sel` with labels for this index coordinates.
+
+        Coordinate label indexers can be of many kinds, e.g., scalar, list,
+        tuple, array-like, slice, :py:class:`Variable`, :py:class:`DataArray`, etc.
+        It is the responsibility of the index to handle those indexers properly.
+
+        Parameters
+        ----------
+        labels : dict
+            A dictionary of coordinate label indexers passed from
+            :py:meth:`Dataset.sel` and where the entries have been filtered
+            for the current index.
+
+        Returns
+        -------
+        sel_results : :py:class:`IndexSelResult`
+            An index query result object that contains dimension positional indexers.
+            It may also contain new indexes, coordinate variables, etc.
+        """
         raise NotImplementedError(f"{self!r} doesn't support label-based selection")
 
     def join(self: T_Index, other: T_Index, how: str = "inner") -> T_Index:
```

Comment:
```suggestion
    def join(self: T_Index, other: T_Index, how: JoinOptions = "inner") -> T_Index:
```
You can use Literals from:
https://github.com/pydata/xarray/blob/19b9e04e9a7f0118f35aa2b8946df0ef7cb3375c/xarray/core/types.py#L64

I'm not so sure about adding typing in the docstring anymore, there's a real risk that they will forgotten about and won't match the typing. 
Dask has a nice guideline with some discussion about it here:
https://github.com/dask/community/issues/255

Reply:
> I'm not so sure about adding typing in the docstring anymore, there's a real risk that they will forgotten about and won't match the typing.

Yes let's remove it. The goal of the Index docstrings is to help those who want to implement a custom index and I assume they are already familiar with Xarray.